Saved/Cached Queries

masojus opened the pull request:
Fixes Issue #87 by adding support for Saved/Cached Queries.
- This satisfies both the MVP and Stretch goals for that issue.
- Support needed to be added in a back-compat way for parsing raw root JSON Arrays in HTTP responses, which wasn't there before. This affects all clients, including Android, so I'll be sure to test this in the sample apps.
- There was no notion of Master Key usage in the Java SDK outside of Scoped Keys, so that was added. If that's not desirable, we could consider scoping the notion of the Master Key to the `SavedQueries` interface, but we'd want to be sure that makes sense for future work (Access Keys and Datasets, for starters) which also need Master Key for some functionality.
- There was no notion of an HTTP verb other than POST before this.
- This change does not invent any special representation for responses to PUT requests or the ancillary data returned from `.../result` endpoints, but rather relies on pre-existing `QueryResult` to represent the actual query result, and just a `Map<String, Object>` to represent everything else like the Saved Query definitions. We could look at adding some helpers if it seems frequent that client code wants to access things like `run_information` and properties related to create/update info.
- Look at individual commit messages for more details.
- I'll follow up with a README/changelog change with new versioning. I was planning to make this be v5.1.0 since it adds a big chunk of functionality.
- I'll open an Issue to fix the `__fake_root` stuff in a better way that requires a breaking change, but in a `vnext_major` branch.
- There are a few remaining TODOs in the code. I'm hoping to get some feedback during PR and address those, or else open Issues for them.

masojus commented on query/src/main/java/io/keen/client/java/KeenQueryClient.java:
```diff
@@ -253,47 +255,125 @@ public QueryResult selectUnique(String eventCollection, String targetProperty, T
+
+    private QueryResult rawMapResponseToQueryResult(KeenQueryRequest request,
+                                                    Map<String, Object> response) {
+        // TODO : Eventually moving result parsing out of this class might be a good plan, like
```
Probably should just create an issue for this one, since it's going to be a large refactor.

baumatron commented on query/src/test/java/io/keen/client/java/PersistentAnalysisTest.java:
```diff
@@ -0,0 +1,554 @@
+    }
+
+    // Make sure result, if provided, is wrapped in root object/array brackets, if not scalar.
+    private String getFakeSavedQueryDefinition(String result, boolean isGroupBy, boolean isInterval, boolean isMultiAnalysis, boolean isFunnel) {
```
Nit: long line.

baumatron commented on query/src/test/java/io/keen/client/java/PersistentAnalysisTest.java:
```diff
@@ -0,0 +1,554 @@
+            analysisType = "sum";
+        }
+
+        String definition = "{" +
```
Oh man... no here string literal in Java? 😞

masojus replied:
Java 👎

baumatron commented on query/src/test/java/io/keen/client/java/PersistentAnalysisTest.java:
```diff
@@ -0,0 +1,554 @@
+        // Should have 'refresh_rate' and 'query' top-level keys, at least.
+        assertTrue("Missing required top-level fields.", 2 <= requestNode.size());
+
+        assertTrue(requestNode.hasNonNull(KeenQueryConstants.REFRESH_RATE));
```
Just to confirm, `refresh_rate` must be specified because if it isn't the API/backend will set it to the minimum cached query interval, right?

masojus replied:
Yes, for now. It seems like a bug, but I'm not sure when they'll get to fixing it, and it's just as easy for us to pass it along.

josephwegner commented on android/src/main/java/io/keen/client/android/AndroidJsonHandler.java:
```diff
@@ -40,8 +44,27 @@
+            // TODO : What should the return value be? Technically it could be a List or Map, so it
+            // should be Object, then client code would need to do the instanceof check. For now,
+            // so as to not break the KeenJsonHandler interface, we can stick a dummy "root" key in
+            // the map we pass back.
```
Is there anywhere in the code where this isn't passed a map? I believe our API will always respond in a map format.

Either way, I think I'm fine with creating a fake root object so it becomes a map... but I'm not terribly opinionated on the matter.

masojus replied:
The problem is that the JSON parser will return either a Map<> or a List<> depending on the root JSON structure. Usually the root is a JSON Object, but now we have to handle a JSON Array as a root object too. So in the `vnext_major` branch we should choose whether to branch here or at a higher layer of abstraction. This TODO is mostly an inquiry to you folks in case you have an opinion as to where that branching should occur, but it doesn't have to be decided today...and if you have no opinion then I'll just do what I think is best in the `vnext_major` branch and leave this as is for the minor release.

josephwegner commented on core/src/main/java/io/keen/client/java/KeenClient.java:
```diff
@@ -1549,13 +1565,19 @@ private void handleSuccess(KeenCallback callback) {
+                                                               keenProperties);
                 }
             } catch (Exception userException) {
                 // Do nothing.
```
This seems like a good question, especially because I think this is calling client code (the success callback). If that errors, the user should know about it.

Not required to fix in this PR, if you don't want to.

masojus replied:
Bummer, I typed up an entire response to this and apparently never clicked submit. Anyway the idea is a telemetry module should *not* take down your app since it's not crucial for even reporting to work in order for the app to work. So, failures are explicitly handled in a callback, or else they're swallowed. I think for event write, this is the proper technique, but for queries I would not do this, because one is asking for data that presumably is needed to show the user something, so failure isn't a tolerable scenario. Especially, in this case, we shouldn't re-throw because it's in the callback to client code. If client code wants to throw an exception in their callback, well, OK, I guess we could opt to never prevent you from shooting yourself in the foot, but I'd say this is a better solution in an event reporting SDK. That said, we should probably log it, so I'll make a note to add logging to these.

masojus replied:
Though @josephwegner makes a good point...if the exception is generated in client code, then perhaps we really *should* re-throw that.

masojus replied:
Created Issue #98.

josephwegner commented on core/src/test/java/io/keen/client/java/TestKeenClientBuilder.java:
```diff
@@ -7,9 +7,8 @@
 public class TestKeenClientBuilder extends KeenClient.Builder {
-
     private final Environment environment;
-    private boolean isNetworkConnected = true;
```
Why remove this?

masojus replied:
I think it just wasn't used, but I'll make sure it has no effect.

josephwegner commented on query/src/main/java/io/keen/client/java/PersistentAnalysis.java:
```diff
@@ -0,0 +1,114 @@
+    // display name of "Can has diacritics Ñüáúéíó üöäñ çàèìòù ãõ" yields a url as follows:
+    // "https://...?saved_query=can-has-diacritics----". So two questions remain:
+    // 1) Can the API handle a queryName with non-ASCII? Should we allow ?
+    // 2) This regex does not match strings like "árbol π" so we only accept ASCII
```
I think disallow, since the API filters them out.

masojus replied:
Ok that's what it does. I'll remove the comment or adjust it to reflect what and why.

josephwegner commented on query/src/main/java/io/keen/client/java/RefreshRate.java:
```diff
@@ -0,0 +1,36 @@
+
+    // The refresh rate range empirically is [14400, 86400] seconds, inclusive at both boundaries.
+    // A refresh rate of 0 means caching is turned off.
+    // TODO : Docs on the website are wrong as of 3/14/17, as has been reported. Get that fixed.
```
Can we remove this TODO? The related changes won't be in this codebase, and either _what_ is wrong is unclear. I have a feeling this will exist forever if we commit it.

masojus replied:
Yeah this TODO will go away no matter what before merging. It's just so you and other reviewers see it and so it's in the git history, because it needs to be fixed but is external to the SDK.

josephwegner commented on query/src/main/java/io/keen/client/java/RefreshRate.java:
```diff
@@ -0,0 +1,36 @@
+    public static final int MAX = 86400; // Maximum is 24 hrs
+
+    public static int fromHours(int hours) {
+        int refreshRate = hours * 3600;
```
nit: can we make a constant called `HOUR_IN_SECONDS`, that contains 3600?

masojus replied:
Yep will do.

josephwegner commented on query/src/main/java/io/keen/client/java/RefreshRate.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    static void validateRefreshRate(int refreshRate) {
+        if (0 != refreshRate && (RefreshRate.MIN > refreshRate || RefreshRate.MAX < refreshRate)) {
```
Also, `NO_CACHING` isn't invalid is it? It just means... no caching?

masojus replied:
Yeah I'll change to `NO_CACHING` and @josephwegner no it isn't invalid, but that's expressed in the `if()` clause.

josephwegner commented on query/src/main/java/io/keen/client/java/SavedQueryPut.java:
```diff
@@ -0,0 +1,111 @@
+                  KeenQueryRequest query,
+                  int refreshRate,
+                  Map<String, ?> miscProperties) {
+        super(HttpMethods.PUT, true /* needsMasterKey */, queryName, displayName);
```
Why the comment here?

masojus replied:
I hate literals as actual parameters when reading code if not in the IDE. In the IDE I can just hover and see what formal parameter that maps to, but reading it elsewhere, like here in Git, it's so much more useful to have the comment for `null` and boolean literals and number type literals, etc. It's something I always do in C and C++ programming, and if there's a good reason not to do it in Java, let me know, but my opinion is that it adds clarity.

baumatron replied:
Sometimes I like to use enums for that reason, just another option. Sometimes that's overkill though. Named parameters would be nice.

masojus replied:
For sure. In .NET or Python I'd just lean on named params. For null and arbitrary numerics that doesn't always help though. I like to either name a var or put the comment there.

masojus commented on query/src/main/java/io/keen/client/java/SavedQueryPut.java:
```diff
@@ -0,0 +1,111 @@
+
+            requestArgs.put(KeenQueryConstants.QUERY, queryArgs);
+
+            // TODO : ... or does it go here? Seems like it works both ways :S
```
None that I've heard.

masojus replied:
@josephwegner Confirmed that it should go on the `query` object as the docs state.

masojus commented on core/src/main/java/io/keen/client/java/KeenProject.java:
```diff
@@ -28,19 +28,35 @@ public KeenProject() {
+     *                  functionality that requires such a key.
+     */
+    public KeenProject(String projectId, String writeKey, String readKey, String masterKey) {
+        if (projectId == null || projectId.trim().isEmpty()) {
```
What sort of validation? Since this is tossed about between modules, we can't yet know if a master is needed or a write or a read, because you may only be doing queries, or only doing administrative operations, or only doing writes, or whatever, so in theory we can't know what's needed. That said, at least one of them should be non-null and non-empty, so at a minimum we could validate that.

masojus commented on java/src/main/java/io/keen/client/java/JacksonJsonHandler.java:
```diff
@@ -25,7 +29,44 @@
-        return mapper.readValue(reader, MAP_TYPE);
+        // TODO : We can't assume the top-level node is a JSON Object anymore, because parts of the
+        // API we access return a JSON Array as the root. So we need to detect the type and decide
+        // what to return, so we need a different return type. Technically it could be a List or
```
Yeah I'll replace this TODO with either a clearer NOTE or just create an Issue, but do you have opinions on how to handle the situation in the next major revision? I'd be happy to hear what layer of abstraction you think should handle the branch between top-level object and top-level array.

masojus commented on query/src/main/java/io/keen/client/java/SavedQueriesImpl.java:
```diff
@@ -0,0 +1,396 @@
+        // We expect a structure such that each entry in the list was a JSON Object representing a
+        // query definition, and no entry should be a JSON Value.
+        for (Object defObj : response) {
+            if (!(defObj instanceof LinkedHashMap)) {
```
I was intentionally more specific only because I happen to know empirically that `LinkedHashMap` is what that lib should produce. But, you're right in that we could be more lenient and future-proof if we accepted the most abstract interface that provides the necessary functionality, so I'll change it. Thanks!

masojus commented on query/src/main/java/io/keen/client/java/RequestUrlBuilder.java:
```diff
@@ -15,43 +16,45 @@
             ));
         } catch (MalformedURLException ex) {
-            Logger.getLogger(RequestUrlBuilder.class.getName()).log(Level.SEVERE, "Failed to format query URL.", ex);
+            Logger.getLogger(RequestUrlBuilder.class.getName()).log(Level.SEVERE,
```
Yeah I still haven't gotten my Android Studio formatting settings totally tuned yet, so sometimes I auto-format and it does weird stuff lik this. I made a note in my TODO list a few weeks ago to assess moving to an `.editorconfig` setup that works cross-IDE and will enforce simple formatting rules similar to the `.clang-format` rules you added. Should probably make that an Issue.

masojus commented on java/src/main/java/io/keen/client/java/JacksonJsonHandler.java:
```diff
@@ -25,7 +29,44 @@
+        Map<String, Object> rootMap = null;
+
+        /*
+        TODO : Are the calls to traverse() here any better or worse than doing either of these?:
```
If nobody has any insight here, I'll just leave it as is.

masojus commented:
I think I've addressed all feedback, either via code changes or by responding here inline or opening an Issue. If this all looks OK, I'll get it merged, and then I'll follow up with a PR that updates the Readme/Changelog and bumps the version to 5.1.0 in preparation for releasing the Maven artifacts.
